pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -111,63 +112,143 @@ func (pl *VolumeZone) Filter(ctx context.Context, _ *framework.CycleState, pod *
+					continue
+				}
+				podPVTopologies = append(podPVTopologies, pvTopology{
+					pvName: pv.Name,
+					key:    key,
+					values: volumeVSet,
+				})
 			}
+		}
+	}
+	return podPVTopologies, nil
+}
 
-			if !volumeVSet.Has(nodeV) {
-				klog.V(10).InfoS("Won't schedule pod onto node due to volume (mismatch on label key)", "pod", klog.KObj(pod), "node", klog.KObj(node), "PV", klog.KRef("", pvName), "PVLabelKey", k)
-				return framework.NewStatus(framework.UnschedulableAndUnresolvable, ErrReasonConflict)
-			}
+// PreFilterExtensions returns prefilter extensions, pod add and remove.
+func (pl *VolumeZone) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter invoked at the filter extension point.
+//
+// It evaluates if a pod can fit due to the volumes it requests, given
+// that some volumes may have zone scheduling constraints.  The requirement is that any
+// volume zone-labels must match the equivalent zone-labels on the node.  It is OK for
+// the node to have more zone-label constraints (for example, a hypothetical replicated
+// volume might allow region-wide access)
+//
+// Currently this is only supported with PersistentVolumeClaims, and looks to the labels
+// only on the bound PersistentVolume.
+//
+// Working with volumes declared inline in the pod specification (i.e. not
+// using a PersistentVolume) is likely to be harder, as it would require
+// determining the zone of a volume during scheduling, and that is likely to
+// require calling out to the cloud provider.  It seems that we are moving away
+// from inline volume declarations anyway.
+func (pl *VolumeZone) Filter(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	// If a pod doesn't have any volume attached to it, the predicate will always be true.
+	// Thus we make a fast path for it, to avoid unnecessary computations in this case.
+	if len(pod.Spec.Volumes) == 0 {
+		return nil
+	}
+	node := nodeInfo.Node()
+	if node == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+	nodeConstraints := make(map[string]string)
+	for _, label := range topologyLabels {
+		v, ok := node.Labels[label]
+		if !ok {
+			continue
+		}
+		nodeConstraints[label] = v
+	}
+	if len(nodeConstraints) == 0 {
+		// The node has no zone constraints, so we're OK to schedule.
+		// In practice, when using zones, all nodes must be labeled with zone labels.
+		// We want to fast-path this case though.
+		return nil
+	}
+
+	var podPVTopologies []pvTopology
+
+	state, err := getStateData(cs)
+	if err != nil {
+		//Fallback to calculate pv list here
+		var status *framework.Status
+		podPVTopologies, status = pl.getPVbyPod(ctx, pod)
+		if !status.IsSuccess() {
+			return status
+		}
+	} else {
+		podPVTopologies = state.podPVTopologies
+	}
+
+	for _, pvTopology := range podPVTopologies {
+		nodeV := nodeConstraints[pvTopology.key]
```

Comment:
We could avoid creating the `nodeConstraints` map and query node.Labels directly.

The check for `len(nodeConstraints) == 0` can be replaced by a boolean `hasAnyNodeConstraint`.

Reply:
I think we can throw away `len(nodeConstraints) == 0` if we query node.Labels directly.
We return nil after iterating `podPVTopologies` , so we only care if node has a constraint that doesn't match pod's constraint. If the node doesn't have any constraint, it'll also return nil here.